harness/determined/common/api/_session.py

```diff
@@ -22,69 +22,22 @@
 )
 
 
-def _do_request(
-    method: str,
-    host: str,
-    path: str,
-    max_retries: Optional[GeneralizedRetry],
-    params: Optional[Dict[str, Any]] = None,
-    json: Any = None,
-    data: Optional[str] = None,
-    headers: Optional[Dict[str, str]] = None,
-    cert: Optional[certs.Cert] = None,
-    timeout: Optional[Union[Tuple, float]] = None,
-    stream: bool = False,
-) -> requests.Response:
-    # Allow the json to come pre-encoded, if we need custom encoding.
-    if json is not None and data is not None:
-        raise ValueError("json and data must not be provided together")
-
-    if json:
-        data = det.util.json_encode(json)
-
-    relpath = path.lstrip("/")
-
-    try:
-        r = det_requests.request(
-            method,
-            f"{host}/{relpath}",
-            params=params,
-            data=data,
-            headers=headers,
-            verify=cert.bundle if cert else None,
-            stream=stream,
-            timeout=timeout,
-            server_hostname=cert.name if cert else None,
-            max_retries=max_retries,
-        )
-    except requests.exceptions.SSLError:
-        raise
-    except requests.exceptions.ConnectionError as e:
-        raise errors.MasterNotFoundException(str(e))
-    except requests.exceptions.RequestException as e:
-        raise errors.BadRequestException(str(e))
-
-    def _get_error_str(r: requests.models.Response) -> str:
-        try:
-            json_resp = _json.loads(r.text)
-            mes = json_resp.get("message")
-            if mes is not None:
-                return str(mes)
-            # Try getting GRPC error description if message does not exist.
-            return str(json_resp.get("error").get("error"))
-        except Exception:
-            return ""
-
-    if r.status_code == 403:
-        raise errors.ForbiddenException(message=_get_error_str(r))
-    if r.status_code == 401:
-        raise errors.UnauthenticatedException()
-    elif r.status_code == 404:
-        raise errors.NotFoundException(_get_error_str(r))
-    elif r.status_code >= 300:
-        raise errors.APIException(r)
-
-    return r
+def _make_requests_session(
```

Comment:
What difference should the reader make between a `requests_session` and an `http_session`?

Reply:
It seems to me like a `BaseSession._http_session` is better named a `_requests_session`, FWIW.

Reply:
ok -- suggestion: replace all the `_make_http_session` methods with calls to `_make_requests_session` that pass explicit values for all these parameters. That way I don't have to cross-reference attributes and figure out what they may have been set to from elsewhere (`with_retry` took a lot of thinking before I felt comfortable that it was correct)

Reply:
Well, now I see that isn't possible. This seems like a smell. Thinking about how to fix it.

Reply:
Well... this would be simpler if, like in requests, the resources were created during `__init__`. In requests.sessions.Session:

```python
    def __enter__(self):
        return self
```

Reply:
Is there any danger in doing it this way? I don't think a connection will be opened inside the requests Session until a GET, etc. is called.

Reply:
well, i tried this at some point. i think the awkwardness here goes back to wanting to support the existing behavior of keeping direct calls single-use sessions. 

but it was my first instinct as well to initialize the session in the __init__. it ended up being more awkward, because then there's non-obvious logic that would be required in `_do_request` on when to close them. 

in the end i felt like it was much more explicit to create resources as the behavior dictates: only on entering a context manager do we start a persistent session, and `_do_request` will use a persistent session if it exists, otherwise it will create a one-time session that only exits for the scope of the function.

Reply:
> because then there's non-obvious logic that would be required in `_do_request` on when to close them.

Could we talk through this? I think it would be nice to know that resources are created rather than have to puzzle out context.